src/ray/gcs/gcs_server/ray_sync.h

```diff
@@ -0,0 +1,114 @@
+// Copyright 2022 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#pragma once
+
+#include <memory>
+#include <type_traits>
+
+#include "ray/common/asio/periodical_runner.h"
+#include "ray/gcs/gcs_server/gcs_resource_report_poller.h"
+#include "ray/gcs/gcs_server/grpc_based_resource_broadcaster.h"
+
+namespace ray {
+namespace sync {
+
+class RaySync {
```

Comment:
the interface looks very clean.

nit: can you separate the implementation into -internal.h and add some comments on the interface?

Reply:
In the comment, could you mention the thread model?

Reply:
@scv119 The interface is still WIP since we need to refactor the code first and migrate it to the final version. It's going to be something similar like (https://github.com/ray-project/ray/pull/22073). Are you OK that I can update it in the later PR (it should be the PR [5])

- [3] Deprecate the old pg reporting code path
- [4] Remove syncer from gcs cluster resource manager and encapsulate everything in syncer module.
- [5] Introduce Reporter/Receiver API in the prototype and adaptor rayler and gcs with that.
- [6] Versioning the report
- [7] Experiment with protocol & communication layer change.

Reply:
Do we have a design doc about the end state once all the PRs are merged? (for example, I'm curious about the reporter/receiver api and versioning)

Reply:
@jjyao this is only refactoring. Prototype is here is you'd like to have a check https://github.com/ray-project/ray/pull/22073

Reply:
oh, got it. I thought it was distributed versioning.